lightweight/calendardataservice/calendardataservice.cpp

```diff
@@ -0,0 +1,102 @@
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE."
+ */
+
+#include "calendardataservice.h"
+
+#include <QtCore/QDate>
+#include <QtCore/QDebug>
+#include <QtDBus/QDBusConnection>
+#include <QtDBus/QDBusMessage>
+
+#include "calendardataserviceadaptor.h"
+#include "../../src/calendaragendamodel.h"
+#include "../../src/calendarevent.h"
+#include "../../src/calendareventoccurrence.h"
+
+CalendarDataService::CalendarDataService(QObject *parent) :
+    QObject(parent), mAgendaModel(0)
+{
+    registerDataTypes();
+    new CalendarDataServiceAdaptor(this);
+    QDBusConnection connection = QDBusConnection::sessionBus();
+    if (!connection.registerService("org.nemomobile.calendardataservice"))
+        qWarning("Can't register org.nemomobile.calendardataservice service on the session bus");
+
+    if (!connection.registerObject("/org/nemomobile/calendardataservice", this))
+        qWarning("Can't register org/nemomobile/calendardataservice object for the D-Bus service.");
+}
+
+void CalendarDataService::getEvents(const QString &startDate, const QString &endDate)
+{
+    QDate start = QDate::fromString(startDate, Qt::ISODate);
+    QDate end = QDate::fromString(endDate, Qt::ISODate);
+    if (!start.isValid() || !end.isValid()) {
+        qWarning() << "Invalid date parameter(s):" << startDate << ", " << endDate;
+        exit(1);
+    }
+    initialize();
+    mAgendaModel->setStartDate(start);
+    mAgendaModel->setEndDate(end);
+}
+
+void CalendarDataService::updated()
+{
+    EventDataList reply;
+    for (int i = 0; i < mAgendaModel->count(); i++) {
+        QVariant variant = mAgendaModel->get(i, NemoCalendarAgendaModel::EventObjectRole);
+        QVariant occurrenceVariant = mAgendaModel->get(i, NemoCalendarAgendaModel::OccurrenceObjectRole);
+        if (variant.canConvert<NemoCalendarEvent *>() && occurrenceVariant.canConvert<NemoCalendarEventOccurrence *>()) {
+            NemoCalendarEvent* event = variant.value<NemoCalendarEvent *>();
+            NemoCalendarEventOccurrence* occurrence = occurrenceVariant.value<NemoCalendarEventOccurrence *>();
+            EventData eventStruct;
+            eventStruct.displayLabel = event->displayLabel();
+            eventStruct.description = event->description();
+            eventStruct.startTime = occurrence->startTime().toString(Qt::ISODate);
+            eventStruct.endTime = occurrence->endTime().toString(Qt::ISODate);
+            eventStruct.allDay = event->allDay();
+            eventStruct.color = event->color();
+            eventStruct.recurrenceId = event->recurrenceIdString();
+            eventStruct.uniqueId = event->uniqueId();
+            reply << eventStruct;
+        }
+    }
+    emit getEventsResult(reply);
+    exit(0);
+}
+
+void CalendarDataService::initialize()
+{
+    if (!mAgendaModel) {
+        mAgendaModel = new NemoCalendarAgendaModel(this);
+        connect(mAgendaModel, SIGNAL(updated()), this, SLOT(updated()));
```

Comment:
Can live with it, but the worker thread doesn't provide too much benefit in this setup. Just directly loading events with mkcal/kcalcore could be quite straightforward.